module/zfs/zpl_file.c

```diff
@@ -361,8 +361,10 @@ zpl_write(struct file *filp, const char __user *buf, size_t len, loff_t *ppos)
 	ssize_t wrote;
 
 	crhold(cr);
+	spl_inode_lock(file_inode(filp));
 	wrote = zpl_write_common(filp->f_mapping->host, buf, len, ppos,
 	    UIO_USERSPACE, filp->f_flags, cr);
+	spl_inode_unlock(file_inode(filp));
```

Comment:
You are right, and sync calls back into fsync which in turn acquires the inode_lock and this will likely lead to a deadlock. Looking around zfs_fsync and zfs_fwrite, it seems the sync part is happening as part of the ZIL commit, so what about moving the locking inside zfs_write and encompassing everything until the zil is about to be committed? Even though ZFS' code is sufficiently different than other upstream filesystems so I'm not entirely sure.